### PR TITLE
Support (un)ignoring users (#546)

### LINKED ISSUE #546: Support ignoring users
Matrix allows ignoring a user: https://matrix.org/docs/spec/client_server/r0.6.1#id144

Nheko should allow ignoring a user as well as show ignored users and allow unignoring them again. There is already a button for that in the user profiles. A /command would be nice too.

### PR DESCRIPTION
**Pull request by NepNep21**
Considerations:
There is currently a minor issue with `TimelineModel::clearTimeline()` where in some seemingly random (I couldn't narrow down the cause) cases it won't attempt to reload the timeline after ignoring an account unless the user clicks out of it and back in. It is also present upstream as seen with `/clear-timeline` so a fix could be discussed here.

It may be a good idea to add commands to (un)ignore a user, (though I don't see much value in it since I imagine most people don't know the server others use off the top of their head and display names are almost always used in the timeline) I will wait for feedback on that idea.

Closes #546

### CODE REVIEW

File: resources/qml/dialogs/IgnoredUsers.qml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                MainWindow.showNotification(text)
+            } else {
+                users = Array.from(users).filter(user => user !== id)
+            }
+        }
+    }
+
+    ListView {
+        id: view
+        width: ignoredUsers.width
```
Comment (deepbluev7) [inline resources/qml/dialogs/IgnoredUsers.qml:37]: usually you do that by using "anchors.fill: parent"

File: resources/qml/dialogs/IgnoredUsers.qml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        height: ignoredUsers.height
+        Layout.leftMargin: Nheko.paddingMedium
+        Layout.rightMargin: Nheko.paddingMedium
```
Comment (deepbluev7) [inline resources/qml/dialogs/IgnoredUsers.qml:40]: These only do something inside a layout, so I don't think they do anything here.

File: resources/qml/dialogs/UserProfile.qml [hunk 3]
```diff
@@ -299,6 +312,25 @@ ApplicationWindow {
                     onClicked: profile.refreshDevices()
                 }
 
+                ImageButton {
```
Comment (deepbluev7) [inline resources/qml/dialogs/UserProfile.qml:315]: Not sure if this is the right place for that button. Maybe the room settings are better, even though ignores are global?

File: src/ui/UserProfile.cpp [hunk 4]
```diff
@@ -224,6 +224,57 @@ UserProfile::refreshDevices()
     fetchDeviceList(this->userid_);
 }
 
+QVector<QString>
+UserProfile::getIgnoredUsers()
```
Comment (deepbluev7) [inline src/ui/UserProfile.cpp:228]: This should probably be a property on the TimelineViewManager. The READ function would return the list of ignored users, while the WRITE function would modify it. the NOTIFY signal would emit a change everytime the account data event is changed.

**Comment by NepNep21** [comment]
There are some questions in the code, I will continue working on this soon.

File: CMakeLists.txt [hunk 5]
```diff
@@ -778,6 +778,7 @@ set(QML_SOURCES
         resources/qml/dialogs/AllowedRoomsSettingsDialog.qml
         resources/qml/dialogs/RoomSettings.qml
         resources/qml/dialogs/UserProfile.qml
+		resources/qml/dialogs/IgnoredUsers.qml
```
Comment (sc0w) [inline CMakeLists.txt:781]: wrong indentation?

**Comment by deepbluev7** [comment]
Do you need any help with the property stuff?

**Comment by NepNep21** [comment]
> Do you need any help with the property stuff?

Not right now at least, just been difficult to debug something else due to #1506, which happens both on my arch host and a fedora VM I made. I think I remember alpine not having that issue so I can try that soon.

File: src/ui/UserProfile.h [hunk 6]
```diff
@@ -184,7 +184,7 @@ class UserProfile final : public QObject
     Q_INVOKABLE void refreshDevices();
     Q_INVOKABLE void banUser();
     Q_INVOKABLE void signOutDevice(const QString &deviceID);
-    // Q_INVOKABLE void ignoreUser();
+    Q_INVOKABLE void ignoredStatus(const QString &id, const bool ignore);
```
Comment (deepbluev7) [inline src/ui/UserProfile.h:187]: Can you make this a Q_PROPERTY? That way you can ignore a user in qml by doing `profile.ignored = true` as well as showing the current status by querying `profile.ignored`.

File: resources/qml/dialogs/UserProfile.qml (hunk 3 above)
Comment (deepbluev7) [reply resources/qml/dialogs/UserProfile.qml:315]: Can you put this into the global settings next to where the button here is: https://github.com/Nheko-Reborn/nheko/pull/1537 ?

File: src/ui/UserProfile.h (hunk 6 above)
Comment (deepbluev7) [reply src/ui/UserProfile.h:187]: Why would a property not be readily available? You can just read that property on the user object as well as write to it, just like you call the function, but you also get proper change notifications.

Comment (NepNep21) [reply src/ui/UserProfile.h:187]: The difference is that, right now, the profile object the ignoredStatus() function is called on isn't necessarily the user being acted on. It *is* when clicking the ignore button in the profile dialog, but not when unignoring the user from the ignored list (see below). 
https://github.com/Nheko-Reborn/nheko/blob/7c9fed91b82c174c40aefdce009090dd142b76f9/resources/qml/dialogs/IgnoredUsers.qml#L65 
I suppose a possible branch of that question would be whether I could easily get a profile object from the user id.

**Comment by deepbluev7** [comment]
Neat, seems to work nicely now! Thanks a lot for your work! I started a pipeline manually now, if that finishes successfully, I will merge it: https://nheko.im/nheko-reborn/nheko/-/pipelines/5185

**Pull request merged**